File: Application/src/main/java/com/teeh/klimasensor/DatabaseService.java
package com.teeh.klimasensor;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.os.AsyncTask;

import com.teeh.klimasensor.common.Constants;
import com.teeh.klimasensor.common.ts.SimpleTs;

import android.util.Log;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import static android.R.attr.id;
import static android.R.id.list;
import static android.media.CamcorderProfile.get;
import static android.webkit.ConsoleMessage.MessageLevel.LOG;

/**
 * Created by teeh on 29.01.2017.
 */


public class DatabaseService {

    private static final String TAG = "DatabaseService";

    private static DatabaseService instance;

    private KlimasensorDbHelper dbHelper;
    private SQLiteDatabase readableDB;
    private SQLiteDatabase writableDB;

    private String[] sensorDataProj = {
            KlimasensorDbContract.KlimasensorEntry._ID,
            KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP,
            KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TEMPERATURE,
            KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE,
            KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_HUMIDITY,
            KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_PRESSURE
    };

    public static DatabaseService getInstance() {
        if (instance == null) {
            instance = new DatabaseService();
        }

        return instance;
    }

    private DatabaseService() {
    }

    ///////////////////////
    // Lifecycle methods //
    ///////////////////////

    public void start(Context context) {
        dbHelper = new KlimasensorDbHelper(context);

        ReadableDBTask readableDBTask = new ReadableDBTask();
        readableDBTask.execute();

        WritableDBTask writableDBTask = new WritableDBTask();
        writableDBTask.execute();
    }

    public void stop() {
        readableDB.close();
        writableDB.close();
        dbHelper.close();
    }

    //////////////////////
    // Database methods //
    //////////////////////

    public Long getNumberOfEntries() {
        return DatabaseUtils.queryNumEntries(readableDB, KlimasensorDbContract.KlimasensorEntry.TABLE_NAME);
    }

    public TsEntry getOldestEntry() {
        return getExtremalEntry("MIN", KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP);
    }

    public TsEntry getLatestEntry() {
        return getExtremalEntry("MAX", KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP);
    }

    public void clearSensorData() {
        writableDB.delete(
                KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                null,
                null
        );
    }

    public List<TsEntry> getAllSensordata() {

        // Filter results WHERE "title" = 'My Title'
        // String selection = FeedEntry.COLUMN_NAME_TITLE + " = ?";
        // String[] selectionArgs = { "My Title" };
        String selection = null;
        String[] selectionArgs = null;

        // How you want the results sorted in the resulting Cursor
        String sortOrder =
                KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " ASC";

        Cursor cursor = readableDB.query(
                KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                sensorDataProj,
                selection,
                selectionArgs,
                null,
                null,
                sortOrder
        );

        Long numEntries = getNumberOfEntries();
        //Long leaveOut = numEntries/1000;
        Long leaveOut = 0L;
        Log.d(TAG, "Leaving out " + leaveOut + " entries.");
        List<TsEntry> result = getTsDataFromCursor(cursor, leaveOut);
        cursor.close();

        return result;
    }

    public List<TsEntry> getAllSensordataRange(Date startDate, Date endDate) {

        Cursor cursor = readableDB.rawQuery("SELECT * FROM " + KlimasensorDbContract.KlimasensorEntry.TABLE_NAME +
                " WHERE " + KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " BETWEEN "+ startDate.getTime() +" AND "+ endDate.getTime() +
                " ORDER BY " + KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " ASC", null);


        Long leaveOut = 0L;
        List<TsEntry> result = getTsDataFromCursor(cursor, leaveOut);
        cursor.close();

        return result;
    }

    public void addNewSensordata(List<String> list) {
        for (String line : list) {
            String[] fields = line.split(",");

            try {
                Date d = SimpleTs.tsFormat.parse(fields[0]);
                Double h = Double.valueOf(fields[1]);
                Double t = Double.valueOf(fields[2]);
                Double p = Double.valueOf(fields[3]);

                ContentValues content = new ContentValues();
                content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP, d.getTime());
                content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_HUMIDITY, h);
                content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TEMPERATURE, t);
                content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_PRESSURE, p);

                long res = writableDB.insert(
                        KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                        null,
                        content
                );


            } catch (NumberFormatException e) {
                Log.e(TAG, e.getLocalizedMessage());
                continue;
            } catch (ParseException e) {
                Log.e(TAG, e.getLocalizedMessage());
                continue;
            }
        }
    }

    public long updateSensordata(List<TsEntry> list) {
        long result = 0;
        for (TsEntry entry : list) {
            ContentValues content = new ContentValues();
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP, entry.getTimestamp().getTime());
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_HUMIDITY, entry.getHumidity());
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TEMPERATURE, entry.getTemperature());
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, entry.getRealTemperature());
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_PRESSURE, entry.getPressure());

            String whereClause = KlimasensorDbContract.KlimasensorEntry._ID + " = ?";
            String whereArgs[] = { entry.getId().toString() };

            long res = writableDB.update(
                    KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                    content,
                    whereClause,
                    whereArgs
            );
            result = result + res;

        }
        return result;
    }

    public boolean updateSensordata(TsEntry entry) {
        List<TsEntry> list = new ArrayList<>();
        list.add(entry);
        long res = updateSensordata(list);
        if (res == 1) {
            return true;
        }
        return false;
    }

    public boolean insertRealTemp(Double value, Integer id) {
        int res = 0;

        if (value != null) {

            ContentValues content = new ContentValues();
            content.put(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, value);

            String whereClause = KlimasensorDbContract.KlimasensorEntry._ID + " = ?";
            String whereArgs[] = { id.toString() };

            res = writableDB.update(
                    KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                    content,
                    whereClause,
                    whereArgs
            );

        }

        if (res == 1) {
            return true;
        }
        else {
            return false;
        }
    }

    /////////////////////
    // Private methods //
    /////////////////////

    private List<TsEntry> getTsDataFromCursor(Cursor cursor) {
        return getTsDataFromCursor(cursor, 0L);
    }

    private List<TsEntry> getTsDataFromCursor(Cursor cursor, Long leaveOut) {
        List<TsEntry> result = new ArrayList<>();
        Long count = 0L;
        while(cursor.moveToNext()) {

            if (leaveOut == 0 || count % leaveOut == 0) {
                Integer id = cursor.getInt(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry._ID));
                Date timestamp = new Date(cursor.getLong(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TIMESTAMP)));
                Double temperature = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_TEMPERATURE));
                Double pressure = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_PRESSURE));
                Double humidity = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_HUMIDITY));
                Double realTemp = null;
                if (!cursor.isNull(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE))) {
                    realTemp = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorDbContract.KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE));
                }

                TsEntry entry = new TsEntry(id, timestamp, humidity, temperature, pressure, realTemp);
                result.add(entry);
            }

            count = count + 1;
        }
        Log.d(TAG, "Read " + result.size() + " entries from cursor");
        return result;
    }

    private TsEntry getExtremalEntry(String extremum, String field) {
        String selection = field + " = (SELECT " + extremum + "(" + field + ") FROM " + KlimasensorDbContract.KlimasensorEntry.TABLE_NAME + ")";

        Cursor cursor = readableDB.query(
                KlimasensorDbContract.KlimasensorEntry.TABLE_NAME,
                sensorDataProj,
                selection,
                null,
                null,
                null,
                null
        );

        List<TsEntry> result = getTsDataFromCursor(cursor);
        cursor.close();

        if (result.size() == 1) {
            return result.get(0);
        }

        if (result.size() > 1) {
            Log.d(TAG, "Extremal entry is not unique!");
        }

        if (result.size() == 0) {
            Log.e(TAG, "No extremal entry found!");
        }

        return null;
    }

    private class ReadableDBTask extends AsyncTask<Void, Void, SQLiteDatabase> {

        protected SQLiteDatabase doInBackground(Void... voids) {
            return dbHelper.getReadableDatabase();
        }

        protected void onPostExecute(SQLiteDatabase result) {
            readableDB = result;
        }
    }

    private class WritableDBTask extends AsyncTask<Void, Void, SQLiteDatabase> {

        protected SQLiteDatabase doInBackground(Void... voids) {
            return dbHelper.getWritableDatabase();
        }

        protected void onPostExecute(SQLiteDatabase result) {
            writableDB = result;
        }
    }
}

File: Application/src/main/java/com/teeh/klimasensor/database/DatabaseService.kt
package com.teeh.klimasensor.database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase

import android.util.Log

import java.text.ParseException
import java.util.ArrayList
import java.util.Date

import com.teeh.klimasensor.TsEntry
import com.teeh.klimasensor.common.activities.BaseActivity
import com.teeh.klimasensor.common.constants.Constants
import com.teeh.klimasensor.common.exception.BusinessException
import com.teeh.klimasensor.common.utils.DateUtils
import com.teeh.klimasensor.rest.SensorData
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.async
import org.jetbrains.anko.doAsync
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class DatabaseService private constructor() {

    private val TAG = "DatabaseService"

    private lateinit var dbHelper: DatabaseHelper
    private lateinit var readableDB: SQLiteDatabase
    private lateinit var writableDB: SQLiteDatabase

    private object Holder { val INSTANCE = DatabaseService() }

    companion object {
        val instance: DatabaseService by lazy { Holder.INSTANCE }

        private val EXTREMUM_MAX: String = "MAX"
        private val EXTREMUM_MIN: String = "MIN"
    }

    private val sensorDataProj = arrayOf(KlimasensorEntry.COLUMN_NAME_ID, KlimasensorEntry.COLUMN_NAME_TIMESTAMP, KlimasensorEntry.COLUMN_NAME_TEMPERATURE, KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, KlimasensorEntry.COLUMN_NAME_HUMIDITY, KlimasensorEntry.COLUMN_NAME_PRESSURE)

    //////////////////////
    // Database methods //
    //////////////////////

    val numberOfEntries: Long
        get() = DatabaseUtils.queryNumEntries(readableDB, KlimasensorEntry.TABLE_NAME)

    val oldestEntry: TsEntry
        get() = getExtremalEntry(EXTREMUM_MIN, KlimasensorEntry.COLUMN_NAME_TIMESTAMP)

    val latestEntry: TsEntry
        get() = getExtremalEntry(EXTREMUM_MAX, KlimasensorEntry.COLUMN_NAME_TIMESTAMP)

    ///////////////////////
    // Lifecycle methods //
    ///////////////////////

    fun start(context: Context) {
        dbHelper = DatabaseHelper.getInstance(context)

        doAsync {
            readableDB = dbHelper.readableDatabase
            writableDB = dbHelper.writableDatabase
        }
    }

    fun stop() {
        readableDB.close()
        writableDB.close()
        dbHelper.close()
    }

    fun getSensordata(leaveOut: Long): List<TsEntry> {
        val selection: String? = null
        val selectionArgs: Array<String>? = null
        val sortOrder = KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " ASC"

        val cursor = readableDB.query(
                KlimasensorEntry.TABLE_NAME,
                sensorDataProj,
                selection,
                selectionArgs, null, null,
                sortOrder
        )

        val result = getTsDataFromCursor(cursor, leaveOut)
        cursor.close()

        return result
    }

    fun getSensordataRange(startDate: LocalDateTime, endDate: LocalDateTime, leaveOut: Long): List<TsEntry> {
        val cursor = readableDB.rawQuery("SELECT * FROM " + KlimasensorEntry.TABLE_NAME +
                " WHERE " + KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " BETWEEN " + DateUtils.toLong(startDate) + " AND " + DateUtils.toLong(endDate) +
                " ORDER BY " + KlimasensorEntry.COLUMN_NAME_TIMESTAMP + " ASC", null)

        val result = getTsDataFromCursor(cursor, leaveOut)
        cursor.close()

        return result
    }

    fun getAllSensordataAsync(): Deferred<List<TsEntry>> {
        return async(CommonPool) {
            getSensordata(0L)
        }
    }

    fun getAllSensordata(): List<TsEntry> {
        return getSensordata(0L)
    }

    fun getAllSensordataRange(startDate: LocalDateTime, endDate: LocalDateTime): List<TsEntry> {
        return getSensordataRange(startDate, endDate, 0L)
    }

    fun addNewSensordata(list: List<String>) {
        for (line in list) {
            val fields = line.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            try {
                val d = DateUtils.toLocalDate(fields[0])
                val h = java.lang.Double.valueOf(fields[1])
                val t = java.lang.Double.valueOf(fields[2])
                val p = java.lang.Double.valueOf(fields[3])

                val content = ContentValues()
                content.put(KlimasensorEntry.COLUMN_NAME_TIMESTAMP, DateUtils.toLong(d))
                content.put(KlimasensorEntry.COLUMN_NAME_HUMIDITY, h)
                content.put(KlimasensorEntry.COLUMN_NAME_TEMPERATURE, t)
                content.put(KlimasensorEntry.COLUMN_NAME_PRESSURE, p)

                val res = writableDB.insert(
                        KlimasensorEntry.TABLE_NAME, null,
                        content
                )


            } catch (e: NumberFormatException) {
                Log.e(TAG, e.localizedMessage)
                continue
            } catch (e: ParseException) {
                Log.e(TAG, e.localizedMessage)
                continue
            }

        }
    }

    fun addNewSensordataFromRest(list: List<SensorData>) {
        for (line in list) {

            try {
                val d = DateUtils.toLocalDate(line.timestamp!!, DateTimeFormatter.RFC_1123_DATE_TIME)
                val h = line.humidity
                val t = line.temperature
                val p = line.pressure

                val content = ContentValues()
                content.put(KlimasensorEntry.COLUMN_NAME_TIMESTAMP, DateUtils.toLong(d))
                content.put(KlimasensorEntry.COLUMN_NAME_HUMIDITY, h)
                content.put(KlimasensorEntry.COLUMN_NAME_TEMPERATURE, t)
                content.put(KlimasensorEntry.COLUMN_NAME_PRESSURE, p)

                val res = writableDB.insert(
                        KlimasensorEntry.TABLE_NAME, null,
                        content
                )


            } catch (e: NumberFormatException) {
                Log.e(TAG, e.localizedMessage)
                continue
            } catch (e: ParseException) {
                Log.e(TAG, e.localizedMessage)
                continue
            }

        }
    }

    fun createEntries(list: List<TsEntry>): Long {
        var result: Long = 0

        for (entry in list) {
            val content = ContentValues()
            content.put(KlimasensorEntry.COLUMN_NAME_TIMESTAMP, DateUtils.toLong(entry.timestamp))
            content.put(KlimasensorEntry.COLUMN_NAME_HUMIDITY, entry.humidity)
            content.put(KlimasensorEntry.COLUMN_NAME_TEMPERATURE, entry.temperature)
            content.put(KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, entry.realTemperature)
            content.put(KlimasensorEntry.COLUMN_NAME_PRESSURE, entry.pressure)

            val res = writableDB.insert(
                    KlimasensorEntry.TABLE_NAME,
                    null,
                    content).toLong()

            result = result + res

        }
        return result
    }

    fun updateEntries(list: List<TsEntry>): Long {
        var result: Long = 0
        for (entry in list) {
            val content = ContentValues()
            content.put(KlimasensorEntry.COLUMN_NAME_TIMESTAMP, DateUtils.toLong(entry.timestamp))
            content.put(KlimasensorEntry.COLUMN_NAME_HUMIDITY, entry.humidity)
            content.put(KlimasensorEntry.COLUMN_NAME_TEMPERATURE, entry.temperature)
            content.put(KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, entry.realTemperature)
            content.put(KlimasensorEntry.COLUMN_NAME_PRESSURE, entry.pressure)

            val whereClause = KlimasensorEntry.COLUMN_NAME_ID + " = ?"
            val whereArgs = arrayOf(entry.id.toString())

            val res = writableDB.update(
                    KlimasensorEntry.TABLE_NAME,
                    content,
                    whereClause,
                    whereArgs
            ).toLong()
            result = result + res

        }
        return result
    }

    fun clearSensorData(): Long {
        return writableDB.delete(
                KlimasensorEntry.TABLE_NAME, "1", null
        ).toLong()
    }

    fun deleteEntries(list: List<TsEntry>): Long {
        var result: Long = 0
        for (entry in list) {
            val whereClause = KlimasensorEntry.COLUMN_NAME_ID + " = ?"
            val whereArgs = arrayOf(entry.id.toString())

            val res = writableDB.delete(
                    KlimasensorEntry.TABLE_NAME,
                    whereClause,
                    whereArgs
            ).toLong()
            result = result + res

        }
        return result
    }

    fun updateEntry(entry: TsEntry): Boolean {
        val list = ArrayList<TsEntry>()
        list.add(entry)
        val res = updateEntries(list)
        return if (res == 1L) {
            true
        } else false
    }

    fun createEntry(entry: TsEntry): Boolean {
        val list = ArrayList<TsEntry>()
        list.add(entry)
        val res = createEntries(list)
        return if (res == -1L) {
            true
        } else false
    }

    fun deleteEntry(entry: TsEntry): Boolean {
        val list = ArrayList<TsEntry>()
        list.add(entry)
        Log.i(BaseActivity.TAG, "delete: deleting" + entry.id)
        val res = deleteEntries(list)
        Log.i(BaseActivity.TAG, "delete: " + res.toString())
        return if (res == 1L) {
            true
        } else false
    }

    fun insertRealTemp(value: Double?, id: Int?): Boolean {
        var res = 0

        if (value != null) {

            val content = ContentValues()
            content.put(KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE, value)

            val whereClause = KlimasensorEntry.COLUMN_NAME_ID + " = ?"
            val whereArgs = arrayOf(id.toString())

            res = writableDB.update(
                    KlimasensorEntry.TABLE_NAME,
                    content,
                    whereClause,
                    whereArgs
            )

        }

        return if (res == 1) {
            true
        } else {
            false
        }
    }

    /////////////////////
    // Private methods //
    /////////////////////

    private fun getTsDataFromCursor(cursor: Cursor, leaveOut: Long = 0L): List<TsEntry> {
        val result = ArrayList<TsEntry>()
        var count: Long = 0L
        while (cursor.moveToNext()) {

            if (leaveOut == 0L || count % leaveOut == 0L) {
                val id = cursor.getInt(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_ID))
                val timestamp = cursor.getLong(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_TIMESTAMP))
                val temperature = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_TEMPERATURE))
                val pressure = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_PRESSURE))
                val humidity = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_HUMIDITY))
                var realTemp: Double? = null
                if (!cursor.isNull(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE))) {
                    realTemp = cursor.getDouble(cursor.getColumnIndexOrThrow(KlimasensorEntry.COLUMN_NAME_REAL_TEMPERATURE))
                }

                val entry = TsEntry(id,
                        DateUtils.toLocalDate(timestamp),
                        humidity,
                        temperature,
                        pressure,
                        realTemp)
                result.add(entry)
            }

            count = count + 1
        }
        Log.d(TAG, "Read ${result.size} entries from cursor (left out $leaveOut)")
        return result
    }

    private fun getExtremalEntry(extremum: String, field: String): TsEntry {
        val selection = field + " = (SELECT " + extremum + "(" + field + ") FROM " + KlimasensorEntry.TABLE_NAME + ")"

        val cursor = readableDB.query(
                KlimasensorEntry.TABLE_NAME,
                sensorDataProj,
                selection, null, null, null, null
        )

        val result = getTsDataFromCursor(cursor)
        cursor.close()

        Log.d(TAG, "Fetched: " + result)

        if (result.size == 1) {
            return result[0]
        }
        else if (result.size > 1) {
            Log.w(TAG, "Extremal entry is not unique!")
            return result[0]
        }
        else {
            Log.w(TAG, "Database is empty.")
            return TsEntry(0, LocalDateTime.of(0, 0, 0, 0, 0), null, null, null)
        }
    }
}
